tests/model/document.js

```diff
@@ -321,42 +321,158 @@ describe( 'Document', () => {
+				writer.insertText( 'foo', docFrag, 0 );
+			} );
+
+			expect( spy.calledOnce ).to.be.false;
+		} );
+	} );
+
+	describe( 'event change:data', () => {
+		it( 'should be fired if there was a change in a document tree in a change block and have a batch as a param', () => {
+			doc.createRoot();
+			const spy = sinon.spy();
+
+			doc.on( 'change:data', ( evt, batch ) => {
+				spy();
+				expect( batch ).to.be.instanceof( Batch );
+			} );
+
+			model.change( writer => {
 				writer.insertText( 'foo', doc.getRoot(), 0 );
 			} );
 
 			expect( spy.calledOnce ).to.be.true;
 		} );
 
-		it( 'should be fired if there was a selection change in an (enqueue)change block', () => {
+		it( 'should be fired before the change event', () => {
 			doc.createRoot();
+
+			const callOrder = [];
+
+			doc.on( 'change:data', () => {
+				callOrder.push( 1 );
+			} );
+
+			doc.on( 'change', () => {
+				callOrder.push( 2 );
+			} );
+
+			model.change( writer => {
+				writer.insertText( 'foo', doc.getRoot(), 0 );
+			} );
+
+			expect( callOrder ).to.deep.equal( [ 1, 2 ] );
+		} );
+
+		it( 'should not be fired if only selection changes', () => {
+			const root = doc.createRoot();
 			const spy = sinon.spy();
 
-			const root = doc.getRoot();
 			root._appendChild( new Text( 'foo' ) );
 
-			doc.on( 'change', spy );
+			doc.on( 'change:data', spy );
 
 			model.change( writer => {
 				writer.setSelection( Range.createFromParentsAndOffsets( root, 2, root, 2 ) );
 			} );
 
-			expect( spy.calledOnce ).to.be.true;
+			sinon.assert.notCalled( spy );
+		} );
+
+		it( 'should be fired if default marker operation is applied', () => {
+			const root = doc.createRoot();
+			const spy = sinon.spy();
+
+			root._appendChild( new Text( 'foo' ) );
+
+			doc.on( 'change:data', spy );
+
+			model.change( writer => {
+				const range = Range.createFromParentsAndOffsets( root, 2, root, 4 );
+				writer.addMarker( 'name', { range, usingOperation: true, affectsData: true } );
+			} );
+
+			sinon.assert.calledOnce( spy );
+		} );
+
+		it( 'should not be fired if the marker operation is applied and marker does not affect data', () => {
+			const root = doc.createRoot();
+			const spy = sinon.spy();
+
+			root._appendChild( new Text( 'foo' ) );
+
+			doc.on( 'change:data', spy );
+
+			model.change( writer => {
+				const range = Range.createFromParentsAndOffsets( root, 2, root, 4 );
+				writer.addMarker( 'name', { range, usingOperation: true } );
+			} );
+
+			sinon.assert.notCalled( spy );
+		} );
+
+		it( 'should be fired if the writer adds marker not managed by using operations', () => {
+			const root = doc.createRoot();
+			const spy = sinon.spy();
+
+			root._appendChild( new Text( 'foo' ) );
+
+			doc.on( 'change:data', spy );
+
+			model.change( writer => {
+				const range = Range.createFromParentsAndOffsets( root, 2, root, 4 );
+				writer.addMarker( 'name', { range, usingOperation: false, affectsData: true } );
+			} );
+
+			sinon.assert.calledOnce( spy );
+		} );
```

Comment:
There should be an integrational test with `updateMarker`:

`addMarker` (`affectsData`: `false`) -> `change` event is fired (ensure it is not `change:data`)
`updateMarker` (`affectsData`: `true`) -> `change:data` event is fired
`updateMarker` (`affectsData`: `false`) -> `change` event is fired (ensure it is not `change:data`)